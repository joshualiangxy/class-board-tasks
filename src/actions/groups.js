import firebase from '../firebase/firebase';
import { v4 as uuid } from 'uuid';
import { addGroup } from './user';

export const newGroup = (name, groupId, module) => ({
  type: 'NEW_GROUP',
  name,
  groupId,
  module
});

// Should return a promise of both adding a user to the group and adding a group to a user.
export const startNewGroup = (groupName, module) => {
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const gid = uuid();
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(gid);

    const userPromise = userRef.update({
      groups: firebase.firestore.FieldValue.arrayUnion(gid)
    });
    const groupPromise = groupRef
      .set({
        name: groupName,
        module
      })
      .catch(error => {
        console.log(error);
      });
    const groupUserPromise = groupRef.collection('users').doc(uid).set({
      uid: uid,
      displayName: getState().user.displayName,
      studentNum: getState().user.studentNum,
      admin: true
    });

    return Promise.all([userPromise, groupPromise, groupUserPromise]).then(
      () => {
        dispatch(newGroup(groupName, gid, module));
        dispatch(addGroup(gid));
      }
    );
  };
};

export const addNewUser = (user, gid) => {
  // Note: No checking on this side, since only valid users can be added by getUser
  // Add user to group
  const groupPromise = firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(user.uid)
    .set({
      displayName: user.displayName,
      studentNum: user.studentNum,
      uid: user.uid,
      admin: false
    });
  // Add group to user's groups array.
  const userPromise = firebase
    .firestore()
    .collection('users')
    .doc(user.uid)
    .update({
      groups: firebase.firestore.FieldValue.arrayUnion(gid)
    });
  return Promise.all([groupPromise, userPromise]);
};

export const setGroups = groups => ({ type: 'SET_GROUPS', groups });

export const startSetGroups = () => {
  const groups = []; // The array of objects to be pushed
  return (dispatch, getState) => {
    firebase
      .firestore()
      .collection('users')
      .doc(getState().auth.user.uid)
      .get()
      .then(userSnapshot => {
        // Get the array of all groups this user is in, and dispatch it to the store.
        const groupIds = userSnapshot.data().groups || []; // Array of group id, or empty array.
        const promises = [];

        groupIds.forEach(gid => {
          promises.push(
            firebase
              .firestore()
              .collection('groups')
              .doc(gid)
              .get()
              .then(snapshot => {
                groups.push({
                  name: snapshot.get('name'),
                  gid,
                  module: snapshot.get('module') || ''
                });
              })
          );
        });

        return Promise.all(promises);
      })
      .then(() => dispatch(setGroups(groups)))
      .catch(error => console.log(error));
  };
};

export const leaveGroup = gid => ({
  type: 'LEAVE_GROUP',
  gid
});

export const startLeaveGroup = (gid, count) => {
  console.log(count);
  return (dispatch, getState) => {
    const uid = getState().auth.user.uid;
    const groupRef = firebase.firestore().collection('groups').doc(gid);

    // Promises
    const userPromise = firebase
      .firestore()
      .collection('users')
      .doc(uid)
      .update({
        groups: firebase.firestore.FieldValue.arrayRemove(gid)
      }); // Remove this group from the user's groups.
    const groupPromise = groupRef.collection('users').doc(uid).delete(); // Remove the user document from this group
    return Promise.all([userPromise, groupPromise])
      .then(() => {
        if (count <= 1) {
          // This is for the case where the user is alone, and only this case, so there shouldn't be
          // any risk of deleting a group someone else is in. Unless someone decides to multi-tab this to break the system.
          // However, this problem can also be solved if we do an additional database-check to ensure there aren't any users left.
          // The count includes the current user.

          // Deletes all of the documents under the users collection.
          const usersPromise = groupRef.collection('users').get().then(snapshot => {
            snapshot.forEach(doc => doc.ref.delete())
          })
            // Deletes all of the documents under the tasks collection.
          const tasksPromise = groupRef.collection('tasks').get().then(snapshot => {
            snapshot.forEach(doc => doc.ref.delete())
          })
          return Promise.all([usersPromise, tasksPromise]).then(() => groupRef.delete());
        } else {
          return;
        }
      })
      .then(() => dispatch(leaveGroup(gid)));
  };
};

export const getUser = email => {
  // Get UID from email
  return firebase
    .firestore()
    .collection('emailToUid')
    .doc(email)
    .get()
    .then(uidDoc => {
      if (uidDoc.exists) {
        // If this user exists, find the UID associated
        return firebase
          .firestore()
          .collection('users')
          .doc(uidDoc.data().uid)
          .get()
          .then(userDoc => {
            // Then return an object containing the user's name and number
            return {
              displayName: userDoc.data().displayName,
              studentNum: userDoc.data().studentNum,
              uid: userDoc.ref.id
            };
          });
      } else {
        return undefined;
      }
    });
};

// Not to be confused with getUser, this one is for all the users in a group.
export const getAllUsers = gid => {
  return () =>
    firebase
      .firestore()
      .collection('groups')
      .doc(gid)
      .collection('users')
      .get()
      .then(query => {
        const result = [];
        query.forEach(doc => {
          result.push(doc.data());
        });
        return result;
      });
};

export const kickUser = (user, gid) => {
  // user is a user object, group is gid
  // Since users aren't stored inside the store, no need to dispatch
  const uid = user.uid;
  // Remove this user from the group's collection of users
  const groupPromise = firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .delete();
  // Remove this group from the user's groups
  const userPromise = firebase
    .firestore()
    .collection('users')
    .doc(uid)
    .update({
      groups: firebase.firestore.FieldValue.arrayRemove(gid)
    });
  // Remove the user from any tasks 
  const tasksPromise = firebase.firestore().collection('tasks').get().then(query => {
    // Not fully implemented yet. 
    // TODO: This should remove the person's field from the completed object of the doc.
    query.forEach(doc => {
      const update = {};

      update[`completed.${uid}`] = firebase.firestore.FieldValue.delete();
      doc.update(update);
    })
  })
  return Promise.all([groupPromise, userPromise]);
};

export const promoteUser = (user, gid) => {
  const uid = user.uid;
  return firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .update({
      admin: true
    });
};
export const demoteUser = (user, gid) => {
  const uid = user.uid;
  return firebase
    .firestore()
    .collection('groups')
    .doc(gid)
    .collection('users')
    .doc(uid)
    .update({
      admin: false
    });
};