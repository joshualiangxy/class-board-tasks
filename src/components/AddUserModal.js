import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
