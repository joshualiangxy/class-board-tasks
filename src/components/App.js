import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

