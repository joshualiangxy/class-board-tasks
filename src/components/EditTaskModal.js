import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
