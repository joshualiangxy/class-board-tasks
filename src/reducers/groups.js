const groupsReducer = (state = [], action) => {
  switch (action.type) {
    case 'NEW_GROUP':
      const newGroup = { name: action.name, gid: action.groupId, module: action.module };
      return [...state, newGroup];
    case 'SET_GROUPS':
      return action.groups;
    case 'LEAVE_GROUP':
      return state.filter(group => group.gid !== action.gid);
    default:
      return state;
  }
};

export default groupsReducer;
