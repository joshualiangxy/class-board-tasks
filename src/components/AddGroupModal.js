import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
