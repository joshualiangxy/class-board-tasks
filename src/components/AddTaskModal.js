import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
