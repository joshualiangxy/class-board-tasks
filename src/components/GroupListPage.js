import React, { useState } from 'react';
import AddGroupModal from './AddGroupModal';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const GroupListPage = ({ groups }) => {
  const [open, setOpen] = useState(false);

  const onRequestClose = () => {
    setOpen(false);
  };

  const openNewGroup = () => {
    setOpen(true);
  };

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Groups</h1>
          <button
            className="button button--norm page-header__actions"
            onClick={openNewGroup}
          >
            Add New Group
          </button>
        </div>
      </div>
      <div className="content-container">
        {
          // Add a 'no groups' when there's no groups
          groups.length === 0 ? (
            <p className="list-header__text">No groups</p>
          ) : (
            groups.map(group => (
              <div key={group.gid} className="grouplist">
                <Link
                  className="button--link-black"
                  to={`/groups/${group.gid}`}
                >
                  <h2 className="groupitem">{group.name}</h2>
                  
                </Link>
              </div>
            ))
          )
        }
      </div>

      <AddGroupModal isOpen={open} onRequestClose={onRequestClose} />
    </div>
  );
};

const mapStateToProps = state => ({ groups: state.groups });

export default connect(mapStateToProps)(GroupListPage);
