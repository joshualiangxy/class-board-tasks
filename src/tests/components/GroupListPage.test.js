import React from 'react';
import { shallow } from 'enzyme';
import { GroupListPage } from '../../components/GroupListPage';
import { noGroups, groupSetOne, groupSetTwo } from '../fixtures/groups';

const wrapper = shallow(<GroupListPage groups={noGroups} />);

describe('render', () => {
  it('should render GroupListPage without groups', () =>
    expect(wrapper).toMatchSnapshot());

  it('should render GroupListPage with one group', () => {
    wrapper.setProps({ groups: groupSetOne });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render GroupListPage with two groups', () => {
    wrapper.setProps({ groups: groupSetTwo });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('addGroupButton', () => {
  it('should have AddGroupModal closed at first', () => {
    expect(wrapper.find("Connect(AddGroupModal)").prop("isOpen")).toBe(false);
  })

  it('should open AddGroupModal on click', () => {
    wrapper.find("button").at(0).prop("onClick")()
    expect(wrapper.find("Connect(AddGroupModal)").prop("isOpen")).toBe(true);
  })

})

describe('addGroupModal', () => {
  it('should close AddGroupModal when calling onRequestClose', () => {
    
    expect(wrapper.find("Connect(AddGroupModal)").prop("isOpen")).toBe(true);
    wrapper.find("Connect(AddGroupModal)").prop("onRequestClose")()
    expect(wrapper.find("Connect(AddGroupModal)").prop("isOpen")).toBe(false);

  })
})