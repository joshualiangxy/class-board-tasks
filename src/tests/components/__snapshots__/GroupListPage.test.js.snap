// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render GroupListPage with one group 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Groups
      </h1>
      <button
        className="button button--norm page-header__actions"
        onClick={[Function]}
      >
        Add New Group
      </button>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="grouplist"
      key="one"
    >
      <Link
        className="button--link-black"
        to="/groups/one"
      >
        <h2
          className="groupitem"
        >
          nameOne
        </h2>
      </Link>
    </div>
  </div>
  <Connect(AddGroupModal)
    isOpen={false}
    onRequestClose={[Function]}
  />
</div>
`;

exports[`render should render GroupListPage with two groups 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Groups
      </h1>
      <button
        className="button button--norm page-header__actions"
        onClick={[Function]}
      >
        Add New Group
      </button>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="grouplist"
      key="one"
    >
      <Link
        className="button--link-black"
        to="/groups/one"
      >
        <h2
          className="groupitem"
        >
          nameOne
        </h2>
      </Link>
    </div>
    <div
      className="grouplist"
      key="two"
    >
      <Link
        className="button--link-black"
        to="/groups/two"
      >
        <h2
          className="groupitem"
        >
          nameTwo
        </h2>
      </Link>
    </div>
  </div>
  <Connect(AddGroupModal)
    isOpen={false}
    onRequestClose={[Function]}
  />
</div>
`;

exports[`render should render GroupListPage without groups 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Groups
      </h1>
      <button
        className="button button--norm page-header__actions"
        onClick={[Function]}
      >
        Add New Group
      </button>
    </div>
  </div>
  <div
    className="content-container"
  >
    <p
      className="list-header__text"
    >
      No groups
    </p>
  </div>
  <Connect(AddGroupModal)
    isOpen={false}
    onRequestClose={[Function]}
  />
</div>
`;
